### PR TITLE
MIG 885 storageclass_conversion

### PR DESCRIPTION
**Pull request by sbeskin-redhat**
MIG 885: storageclass conversion feature
https://issues.redhat.com/browse/MIG-885

1.7
OpenShift 4.9 and higher

Preview:
https://deploy-preview-42719--osdocs.netlify.app/openshift-enterprise/latest/migration_toolkit_for_containers/advanced-migration-options-mtc.html#migration-converting-storage-classes_advanced-migration-options-mtc

### CODE REVIEW

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Module included in the following assemblies:
+//
+// * migration_toolkit_for_containers/advanced-migration-options-mtc.adoc
+
+:_content-type: PROCEDURE
+[id="migration-converting-the-storage-class-of-persistent-volumes-cam_{context}"]
+= Converting the storage class of persistent volumes in the {mtc-short} web console
+
+You can convert the storage class of persistent volumes (PVs) by migrating them within the same cluster. For that, you must create and run a migration plan in the {mtc-full} ({mtc-short}) web console.
```
Comment (stoobie) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:9]:
```suggestion
You can convert the storage class of persistent volumes (PVs) by migrating them within the same cluster. To do so, you must create and run a migration plan in the {mtc-full} ({mtc-short}) web console.
```

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+* You must be logged in as a user with `cluster-admin` privileges on the cluster on which {mtc-short} is running.
+* You must add the cluster to the {mtc-short} web console.
+
+.Procedure
+
+. In the left-side navigation pane of the web console, click *Projects*.
+. In the list of projects, click your project.
++
+The *Project details* page opens.
+. Click the DeploymentConfig. Note the name of its running pod.
```
Comment (stoobie) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:22]: > Click the DeploymentConfig.

Suggestion: Click the DeploymentConfig button.

Avoid using UI elements as words.

**Review by stoobie — CHANGES_REQUESTED** [review]
Looks good! I made a couple of minor comments.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+. Open the YAML tab of the project. Find the PVs and note the names of their corresponding persistent volume claims (PVCs).
+. In the {mtc-short} web console, click *Migration plans*.
+. Click *Add migration plan*.
+. Enter the *Plan name*.
++
+The migration plan name must not exceed 253 lower-case alphanumeric characters (`a-z, 0-9`) and must not contain spaces or underscores (`_`).
+
+. In *Migration type*, select *Storage class conversion*.
+. In *Source cluster*, select *host*.
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:31]: In *Source cluster* , select the desired cluster for storage class conversion.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+. Click *Next*.
++
+The *Namespaces* page opens.
+. Select the required project.
+. Click *Next*.
++
+The *Persistent volumes* page opens. The page displays the PVs in the project, all selected by default.
+. For each PV, select the desired target storage class.
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:39]: target storage class is a field name.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+. Click *Next*.
++
+The wizard validates the new migration plan and shows that it is ready.
+. Click *Close*.
++
+The new plan appears on the *Migration plans* page.
+. To start the conversion, click the three dots menu of the new plan and select *Migrations*, then *Cutover*.
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:46]: To start the conversion, click the three dots menu of the new plan and select *Cutover* option visible under the Migrations.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
++
+The *Cutover migration* notification appears.
+. Click *Migrate*.
++
+The *Cutover started* message appears.
+.  To see the status of the current migration, click the number in the *Migration* column.
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:52]: ** Migrations ** column

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
++
+The *Migrations* page opens.
+. To see more details on the current migration and monitor its progress, click the *Type* column.
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:55]: To see more details on the current migration and monitor its progress, choose the migration from the *Type* column.

Reason for suggested change: There can be multiple migrations for a single migplan.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 1 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:9]: Done

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 2 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:22]: Changed

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+. In the list of projects, click your project.
++
+The *Project details* page opens.
+. Click the DeploymentConfig. Note the name of its running pod.
+. Open the YAML tab of the project. Find the PVs and note the names of their corresponding persistent volume claims (PVCs).
+. In the {mtc-short} web console, click *Migration plans*.
+. Click *Add migration plan*.
+. Enter the *Plan name*.
++
+The migration plan name must not exceed 253 lower-case alphanumeric characters (`a-z, 0-9`) and must not contain spaces or underscores (`_`).
```
[the remark this answers is not in the document]
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:28]: Done

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 3 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:31]: @stillalearner  Mike Turley wrote this: 
"...select "host" as our source cluster (the "host" cluster is automatically configured when MTC is installed, it is the cluster MTC is running on)"
Is it wrong?

Comment (stillalearner) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:31]: @sbeskin-redhat  the storageclass conversion can be performed on any of the clusters (source or target) added in the UI, not necessary the host only.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 4 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:39]: @stillalearner  I mean one of the possible menu options here, not the name of the field.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 5 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:46]: Changed to "select *Migrations* -> *Cutover*"

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 7 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:55]: Done

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+You can convert the storage class of persistent volumes (PVs) by migrating them within the same cluster. For that, you must create and run a migration plan in the {mtc-full} ({mtc-short}) web console.
+
+.Prerequisites
+
+* You must be logged in as a user with `cluster-admin` privileges on the cluster on which {mtc-short} is running.
+* You must add the cluster to the {mtc-short} web console.
+
+.Procedure
+
+. In the left-side navigation pane of the web console, click *Projects*.
```
[the remark this answers is not in the document]
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:18]: OCP added

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 3 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:31]: Changed

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 6 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:52]: Corrected

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
++
+The *Cutover started* message appears.
+.  To see the status of the current migration, click the number in the *Migration* column.
++
+The *Migrations* page opens.
+. To see more details on the current migration and monitor its progress, click the *Type* column.
++
+The *Migration details* page opens.
+When the migration progresses to the DirectVolume step and the status of the step becomes `Running Rsync Pods to migrate Persistent Volume data`, you can click *View details* and see the detailed status of the copies.
+. Open the *PersistentVolumeClaims* tab.
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:59]: There is no PersistentVolumeClaims tab in mtc console, is here also ocp console need to be mentioned ?

Comment (stillalearner) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:59]: Its better to keep both the ocp console steps to check PVC and Pods at the end after completion of migration.

Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc]: @stillalearner  Yes, I've added OCP console.
Agreed, changed

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 5 above)
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc]: So, actually there are 2 options

Stage & Cutover

Shall we mention the description of both and ask the user to choose appropriately?

Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc]: @stillalearner Is the Stage option relevant for storageclass conversion?

Comment (stillalearner) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc]: @sbeskin-redhat  yes, it is. Here are the details of both migrations. Stage and Cutover

During a stage migration:
PV data is copied to the converted PVs.
PVC references in the applications are **not** updated to new PVCs.

During a cutover migration:
All transactions on source applications are halted for the duration of the migration.
PVC references in the applications are updated to new PVCs before restarting the applications.

Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc]: @stillalearner  So what are the considerations for choosing one of those options? And if the user chooses Stage, how does the migration proceeds?

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+The *Persistent volumes* page opens. The page displays the PVs in the project, all selected by default.
+. For each PV, select the desired target storage class.
+. Click *Next*.
++
+The wizard validates the new migration plan and shows that it is ready.
+. Click *Close*.
++
+The new plan appears on the *Migration plans* page.
+. To start the conversion, click the three dots menu of the new plan and select *Migrations*.
++
```
Comment (stillalearner) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:47]: In the three dots menu, "Migrations" is not a selectable object, It is just a heading (plain text) which is always there and under which the Stage and Cutover are the buttons.

Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:47]: Corrected

**Review by stillalearner — APPROVED** [review]

File: migration_toolkit_for_containers/advanced-migration-options-mtc.adoc [hunk 12]
```diff
@@ -43,6 +43,7 @@ include::modules/migration-mapping-destination-namespaces-in-the-migplan-cr.adoc
 include::modules/migration-excluding-pvcs.adoc[leveloffset=+2]
 include::modules/migration-mapping-pvcs.adoc[leveloffset=+2]
 include::modules/migration-editing-pvs-in-migplan.adoc[leveloffset=+2]
+include::modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc[leveloffset=+2]
```
Comment (apinnick) [inline migration_toolkit_for_containers/advanced-migration-options-mtc.adoc:46]: Location looks OK. File name should not contain 'cam'. I also think "migration-converting-storage-classes.adoc" would be a better file name.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 13]
```diff
@@ -0,0 +1,73 @@
+// Module included in the following assemblies:
+//
+// * migration_toolkit_for_containers/advanced-migration-options-mtc.adoc
+
+:_content-type: PROCEDURE
+[id="migration-converting-the-storage-class-of-persistent-volumes-cam_{context}"]
+= Converting the storage class of persistent volumes in the {mtc-short} web console
```
Comment (apinnick) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:7]: Suggest changing title to "Converting storage classes".

If a CLI procedure exists, it should be documented.

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+You can convert the storage class of persistent volumes (PVs) by migrating them within the same cluster. To do so, you must create and run a migration plan in the {mtc-full} ({mtc-short}) web console.
```
Comment (apinnick) [inline modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc:9]: Either "storage classes of persistent volumes" or "the storage class of a persistent volume"

**Review by GroceryBoyJr — COMMENTED** [review]
/LGTM 
Please note: I am in training. Stand by for backup review.

**Comment by ahardin-rh** [comment]
Preview build: https://deploy-preview-42719--osdocs.netlify.app/openshift-enterprise/latest/migration_toolkit_for_containers/advanced-migration-options-mtc.html#migration-converting-storage-classes_advanced-migration-options-mtc

In the future, please provide the full URL at the top of your pull request for the convenience of your reviewers. Thanks! :smile:

**Comment by ahardin-rh** [comment]
@sbeskin-redhat A few more things before we can move this PR forward:

- Can you please provide a direct link to the JIRA /BZ/ Issue that corresponds to this request for tracking purposes? 
- Can you please indicate what OpenShift versions this change applies to?
- Can you please squash all commits down to 1 commit?
-  We will need QE verification before we merge.

Your edits look good otherwise! Thanks again!

File: modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc (hunk 11 above)
Comment (sbeskin-redhat) [reply modules/migration-converting-the-storage-class-of-persistent-volumes-cam.adoc]: @stillalearner 
Sachin, please approve this PR. Thank you!

**Review by stillalearner — APPROVED** [review]
/lgtm

**Comment by stoobie** [comment]
@ahardin-rh MTC 1.7 is now live. Please merge this.

**Pull request merged**